src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -998,4 +1036,294 @@ private static MethodHandle stringValueOf(Class<?> ptype) {
     private StringConcatFactory() {
         // no instantiation
     }
+
+    /**
+     * Simplified concatenation method to facilitate {@link StringTemplate}
+     * concatenation. This method returns a single concatenation method that
+     * interleaves fragments and values. fragment|value|fragment|value|...|value|fragment.
+     * The number of fragments must be one more that the number of ptypes.
+     * The total number of slots used by the ptypes must be less than or equal
+     * to MAX_INDY_CONCAT_ARG_SLOTS.
+     *
+     * @param fragments list of string fragments
+     * @param ptypes    list of expression types
+     *
+     * @return {@link MethodHandle}
+     *
+     * @throws StringConcatException If any of the linkage invariants are violated.
+     * @throws NullPointerException If any of the incoming arguments is null.
+     *
+     * @since 20
+     */
+    @PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+    public static MethodHandle makeConcatWithTemplate(
+            List<String> fragments,
+            List<Class<?>> ptypes)
+            throws StringConcatException
+    {
+        Objects.requireNonNull(fragments, "fragments is null");
+        Objects.requireNonNull(ptypes, "ptypes is null");
+        ptypes = List.copyOf(ptypes);
+
+        if (fragments.size() != ptypes.size() + 1) {
+            throw new StringConcatException("fragments size not equal ptypes size plus one");
+        }
+
+        if (ptypes.isEmpty()) {
+            return MethodHandles.constant(String.class, fragments.get(0));
+        }
+
+        Class<?>[] ttypes = new Class<?>[ptypes.size()];
+        MethodHandle[] filters = new MethodHandle[ptypes.size()];
+        int slots = 0;
+
+        int pos = 0;
+        for (Class<?> ptype : ptypes) {
+            slots += ptype == long.class || ptype == double.class ? 2 : 1;
+
+            if (MAX_INDY_CONCAT_ARG_SLOTS < slots) {
+                throw new StringConcatException("Too many concat argument slots: " +
+                        slots + ", can only accept " + MAX_INDY_CONCAT_ARG_SLOTS);
+            }
+
+            boolean isSpecialized = ptype.isPrimitive();
+            boolean isFormatConcatItem = FormatConcatItem.class.isAssignableFrom(ptype);
+            Class<?> ttype = isSpecialized ? promoteToIntType(ptype) :
+                             isFormatConcatItem ? FormatConcatItem.class : Object.class;
+            MethodHandle filter = isFormatConcatItem ? null : stringifierFor(ttype);
+
+            if (filter != null) {
+                filters[pos] = filter;
+                ttype = String.class;
+            }
+
+            ttypes[pos++] = ttype;
+        }
+
+        MethodHandle mh = MethodHandles.dropArguments(newString(), 2, ttypes);
```

Comment:
The code here us pretty dense. I suggest adding comments that show the type of `mh` as far as it is known. I think it makes it a much easier to keep track of what all the argument indexes mean. e.g.
```suggestion
        // (byte[],long,ttypes...) -> String
        MethodHandle mh = MethodHandles.dropArguments(newString(), 2, ttypes);
```

Reply:
Noted